File: modules/api/src/main/java/net/countercraft/movecraft/events/CraftReleaseEvent.java
package net.countercraft.movecraft.events;

import net.countercraft.movecraft.craft.Craft;
import org.bukkit.event.Cancellable;
import org.bukkit.event.HandlerList;
import org.jetbrains.annotations.NotNull;

/**
 * Called whenever a craft is released
 * @see Craft
 */
@SuppressWarnings("unused")
public class CraftReleaseEvent extends CraftEvent implements Cancellable {
    @NotNull private final Reason reason;
    private static final HandlerList HANDLERS = new HandlerList();
    private boolean cancelled = false;

    public CraftReleaseEvent(@NotNull Craft craft, @NotNull Reason reason) {
        super(craft);
        this.reason = reason;
    }

    @NotNull
    public Reason getReason() {
        return reason;
    }

    public enum Reason{
        DISCONNECT,SUB_CRAFT,PLAYER,FORCE,EMPTY,SUNK,REPAIR,DEATH,INACTIVE
    }

    @Override
    public boolean isCancelled() {
        return this.cancelled;
    }

    @Override
    public void setCancelled(boolean cancel) {
        this.cancelled = cancel;
    }

    @Override
    public HandlerList getHandlers() {
        return HANDLERS;
    }

    @SuppressWarnings("unused")
    public static HandlerList getHandlerList() {
        return HANDLERS;
    }
}

File: src/main/kotlin/dev/cccm5/movecraft/event/CraftReleaseEvent.kt
package dev.cccm5.movecraft.event

import dev.cccm5.movecraft.CraftStateManager
import dev.cccm5.movecraft.craft.PlayerCraft
import org.bukkit.event.HandlerList
import org.bukkit.block.Block

/**
 * Invoked whenever a [PlayerCraft] is removed from the CraftManager
 */
class CraftReleaseEvent(craft: PlayerCraft, val reason: Reason): CraftRemoveEvent(craft) {
    override fun getHandlers(): HandlerList {
        return handlerList
    }

    companion object {
        @JvmStatic private val handlerList = HandlerList()
        @JvmStatic fun getHandlerList() = handlerList
    }

    enum class Reason{
        DISCONNECT, SINK, MANUAL
    }
}